src/core/border_router/nat64_translator.cpp

```diff
@@ -0,0 +1,454 @@
+{
+    LogInfo("mapping removed: %s -> %s", aMapping.mIp6.ToString().AsCString(), aMapping.mIp4.ToString().AsCString());
+    mIp4AddressPool[mAvailableAddressCount] = aMapping.mIp4;
+    mAvailableAddressCount++;
+    mAddressMappingPool.Free(aMapping);
+}
+
+Nat64Translator::AddressMapping *Nat64Translator::CreateMapping(const Ip6::Address &aAddr)
+{
+    AddressMapping *mapping = mAddressMappingPool.Allocate();
+
+    if (mapping == nullptr)
+    {
+        uint64_t                   now = Get<Uptime>().GetUptime();
+        LinkedList<AddressMapping> idleMappings;
+
+        mActiveAddressMappings.RemoveAllMatching(now, idleMappings);
+
+        for (AddressMapping *idleMapping = idleMappings.Pop(); idleMapping != nullptr; idleMapping = idleMappings.Pop())
+        {
+            ReleaseMapping(*idleMapping);
+        }
+
+        mapping = mAddressMappingPool.Allocate();
+    }
+
+    if (mAvailableAddressCount == 0)
+    {
+        if (mapping != nullptr)
+        {
+            mAddressMappingPool.Free(*mapping);
+            mapping = nullptr;
+        }
+        ExitNow();
+    }
+
+    VerifyOrExit(mapping != nullptr);
+
+    mapping->mIp6 = aAddr;
+    mAvailableAddressCount--;
+    mapping->mIp4 = mIp4AddressPool[mAvailableAddressCount];
+    mapping->Touch(Get<Uptime>().GetUptime());
+    mActiveAddressMappings.Push(*mapping);
+    LogInfo("mapping created: %s -> %s", mapping->mIp6.ToString().AsCString(), mapping->mIp4.ToString().AsCString());
+
+exit:
+    return mapping;
+}
+
+Nat64Translator::AddressMapping *Nat64Translator::GetMapping(const Ip6::Address &aAddr, bool aTryCreate)
+{
+    AddressMapping *mapping = mActiveAddressMappings.FindMatching(aAddr);
+
+    // Exit if we found a valid mapping.
+    VerifyOrExit(mapping == nullptr);
+    // If we don't have a valid mapping and we don't want to create one.
+    VerifyOrExit(aTryCreate);
+
+    mapping = CreateMapping(aAddr);
+
+exit:
+    return mapping;
+}
+
+Nat64Translator::AddressMapping *Nat64Translator::GetMapping(const Ip4::Address &aAddr)
+{
+    AddressMapping *mapping = mActiveAddressMappings.FindMatching(aAddr);
+
+    if (mapping != nullptr)
+    {
+        mapping->Touch(Get<Uptime>().GetUptime());
+    }
+    return mapping;
+}
+
+Error Nat64Translator::TranslateIcmp4(AddressMapping &aMapping, Message &aMessage)
+{
+    Error             err = kErrorNone;
+    Ip4::Icmp::Header icmp4Header;
+    Ip6::Icmp::Header icmp6Header;
+
+    OT_UNUSED_VARIABLE(aMapping);
+
+    VerifyOrExit(aMessage.ReadBytes(0, &icmp4Header, sizeof(icmp4Header)) == sizeof(icmp4Header), err = kErrorParse);
+    switch (icmp4Header.GetType())
+    {
+    case Ip4::Icmp::Header::Type::kTypeEchoReply:
+    {
+        VerifyOrExit(aMessage.ReadBytes(0, &icmp6Header, sizeof(icmp6Header)) == sizeof(icmp6Header),
+                     err = kErrorParse);
+        icmp6Header.SetType(Ip6::Icmp::Header::Type::kTypeEchoReply);
+        aMessage.WriteBytes(0, &icmp6Header, sizeof(icmp6Header));
```

Comment:
`aMessage.Write(0, icmp6Header)`

Can use `Message::Write<ObjectType>()` for this.